File: app/src/main/java/com/nfcalarmclock/maxsnooze/NacMaxSnoozePreference.java
package com.nfcalarmclock.maxsnooze;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import androidx.fragment.app.FragmentManager;
import androidx.preference.Preference;

import com.nfcalarmclock.dialog.NacScrollablePickerDialogFragment;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedDefaults;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.R;

/**
 * Preference that displays the max number of snoozes for an alarm.
 */
public class NacMaxSnoozePreference
	extends Preference
	implements NacScrollablePickerDialogFragment.OnScrollablePickerOptionSelectedListener
{

	/**
	 * Preference value.
	 */
	private int mValue;

	/**
	 */
	public NacMaxSnoozePreference(Context context)
	{
		this(context, null);
	}

	/**
	 */
	public NacMaxSnoozePreference(Context context, AttributeSet attrs)
	{
		this(context, attrs, 0);
	}

	/**
	 */
	public NacMaxSnoozePreference(Context context, AttributeSet attrs,
		int style)
	{
		super(context, attrs, style);
		setLayoutResource(R.layout.nac_preference);
	}

	/**
	 */
	@Override
	public CharSequence getSummary()
	{
		NacSharedConstants cons = new NacSharedConstants(getContext());
		int value = this.getValue();

		return NacSharedPreferences.getMaxSnoozeSummary(cons, value);
	}

	/**
	 * Get the preference value.
	 *
	 * @return The preference value.
	 */
	public int getValue()
	{
		return this.mValue;
	}

	/**
	 * @return The default value.
	 */
	@Override
	protected Object onGetDefaultValue(TypedArray a, int index)
	{
		NacSharedDefaults defs = new NacSharedDefaults(getContext());
		return a.getInteger(index, defs.getMaxSnoozeIndex());
	}

	/**
	 * Save the selected value from the scrollable picker.
	 */
	@Override
	public void onScrollablePickerOptionSelected(int index)
	{
		this.mValue = index;

		persistInt(index);
		notifyChanged();
	}

	/**
	 * Set the initial preference value.
	 */
	@Override
	protected void onSetInitialValue(Object defaultValue)
	{
		if (defaultValue == null)
		{
			this.mValue = getPersistedInt(this.mValue);
		}
		else
		{
			this.mValue = (Integer) defaultValue;
			persistInt(this.mValue);
		}
	}

	/**
	 * Show the auto dismiss dialog.
	 */
	public void showDialog(FragmentManager manager)
	{
		NacMaxSnoozeDialog dialog = new NacMaxSnoozeDialog();
		int value = this.getValue();

		dialog.setDefaultScrollablePickerIndex(value);
		dialog.setOnScrollablePickerOptionSelectedListener(this);
		dialog.show(manager, NacMaxSnoozeDialog.TAG);
	}

}

File: app/src/main/java/com/nfcalarmclock/maxsnooze/NacMaxSnoozePreference.kt
package com.nfcalarmclock.maxsnooze

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.fragment.app.FragmentManager
import androidx.preference.Preference
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedConstants
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.view.dialog.NacScrollablePickerDialogFragment.OnScrollablePickerOptionSelectedListener

/**
 * Preference that displays the max number of snoozes for an alarm.
 */
class NacMaxSnoozePreference @JvmOverloads constructor(

	/**
	 * Context.
	 */
	context: Context,

	/**
	 * Attribute set.
	 */
	attrs: AttributeSet? = null,

	/**
	 * Default style.
	 */
	style: Int = 0

	// Constructor
) : Preference(context, attrs, style),

	// Interface
	OnScrollablePickerOptionSelectedListener
{

	/**
	 * Max snooze index.
	 */
	private var maxSnoozeIndex = 0

	/**
	 * Constructor.
	 */
	init
	{
		layoutResource = R.layout.nac_preference
	}

	/**
	 * Get the summary text.
	 *
	 * @return The summary text.
	 */
	override fun getSummary(): CharSequence?
	{
		val cons = NacSharedConstants(context)

		return NacSharedPreferences.getMaxSnoozeSummary(cons, maxSnoozeIndex)
	}

	/**
	 * Get the default value.
	 *
	 * @return The default value.
	 */
	override fun onGetDefaultValue(a: TypedArray, index: Int): Any
	{
		val defaultValue = context.resources.getInteger(R.integer.default_max_snooze_index)

		return a.getInteger(index, defaultValue)
	}

	/**
	 * Save the selected value from the scrollable picker.
	 */
	override fun onScrollablePickerOptionSelected(index: Int)
	{
		// Set the new max snooze index
		maxSnoozeIndex = index

		// Persist the index
		persistInt(index)

		// Notify of a change
		notifyChanged()
	}

	/**
	 * Set the initial preference value.
	 */
	override fun onSetInitialValue(defaultValue: Any?)
	{
		// Check if the default value is null
		if (defaultValue == null)
		{
			maxSnoozeIndex = getPersistedInt(maxSnoozeIndex)
		}
		// Convert the default value
		else
		{
			maxSnoozeIndex = defaultValue as Int

			persistInt(maxSnoozeIndex)
		}
	}

	/**
	 * Show the auto dismiss dialog.
	 */
	fun showDialog(manager: FragmentManager)
	{
		// Create the dialog
		val dialog = NacMaxSnoozeDialog()

		// Setup the dialog
		dialog.defaultScrollablePickerIndex = maxSnoozeIndex
		dialog.onScrollablePickerOptionSelectedListener = this

		// Show the dialog
		dialog.show(manager, NacMaxSnoozeDialog.TAG)
	}

}